hera_pspec/grouping.py

```diff
@@ -288,74 +307,135 @@ def average_spectra(uvp_in, blpair_groups=None, time_avg=False,
                     ints = uvp.get_integrations((spw, blp, p))[:, None]
-                    w = (ints * np.sqrt(nsmp))
+                    # shape of ints: (Ntimes, 1)
                     if store_cov:
                         cov = uvp.get_cov((spw, blp, p))
-                    # Get error bar weights and set invalid ones to zero.
-                    errws = {}
-                    for stat in stat_l:
-                        errs = uvp.get_stats(stat, (spw, blp, p)).copy()
-                        no_data = np.isnan(errs)
-                        errs[~no_data] = 1. / errs[~no_data] ** 2.
-                        errs[no_data] = 0.
-                        errws[stat] = errs
-
-                    # Take time average if desired
-                    if time_avg:
-                      data = (np.sum(data * w, axis=0) \
-                           / np.sum(w, axis=0).clip(1e-10, np.inf))[None]
-                      wgts = (np.sum(wgts * w[:, None], axis=0) \
-                           / np.sum(w, axis=0).clip(1e-10, np.inf)[:, None])[None]
-                      ints = (np.sum(ints * w, axis=0) \
-                           / np.sum(w, axis=0).clip(1e-10, np.inf))[None]
-                      nsmp = np.sum(nsmp, axis=0)[None]
-                      w = np.sum(w, axis=0)[None]
-                      if store_cov:
-                          cov = (np.sum(cov * w[:, None], axis=0) \
-                          / np.sum(w,axis=0).clip(1e-10, np.inf))[None]
-
-                      for stat in stat_l:
-                        errws[stat] = np.sum(errws[stat], axis=0)[None]
-                    # Add multiple copies of data for each baseline according
-                    # to the weighting/multiplicity
-                    for m in range(int(blpg_wgts[k])):
-                        bpg_data.append(data * w)
-                        bpg_wgts.append(wgts * w[:, None])
-                        bpg_ints.append(ints * w)
-                        bpg_nsmp.append(nsmp)
-                        [bpg_stats[stat].append(errws[stat]) for stat in stat_l]
-                        w_list.append(w)
-                        if store_cov:
-                            bpg_cov.append(cov * w[:, None])
 
+                    if if_error_weighted:
+                    # If if_error_weighted==True, all arrays are weighted by a specified kind of errors.
+                    # For each power spectrum P_i with error sigma_i, 
+                    # P_avg = \sum{ P_i / (sigma_i)^2 } / \sum{ 1 / (sigma_i)^2 } 
+                    # while for terms like variance or covariance,
+                    # epsilon_avg = \sum{ (epsilon_i / (sigma_i)^4 } / ( \sum{ 1 / (sigma_i)^2 } )^2
+                    # For reference: Tegmark 1997, The Astrophysical Journal Letters, 480, L87, Table 1, #3
+                    # or Dillon 2014, Physical Review D, 89, 023002 , Equation 34. 
+                        error_weights = uvp.get_stats(name_error_weights, (spw, blp, p)).copy()
```

Comment:
Perhaps these two lines can be merged into one, so `error_weights` is the right quantity immediately (then there would be no need for the `.copy()` either).

Reply:
Sure.